File: packages/cli/src/commands/src.ts
import process from 'node:process'
import { defineCommand } from 'citty'
import { z } from 'zod'
import { ensureCheckout as defaultEnsureCheckout, NoCacheError } from './ensure-checkout.js'

export interface RunSrcOptions {
  spec: string
  projectDir: string
  noFetch?: boolean
  json?: boolean
}

// ── JSON output model ─────────────────────────────────────────────
// Schema for `ask src <spec> --json`. This is the stable machine-
// readable handoff that downstream tools (e.g. csp — code-search)
// consume: `checkoutDir` is the version-pinned, content-stable store
// path they index/search. Mirrors the `DocsModelSchema` precedent in
// docs.ts (zod-schema-as-output-contract).

export const SrcModelSchema = z.object({
  spec: z.string(),
  owner: z.string(),
  repo: z.string(),
  ref: z.string(),
  resolvedVersion: z.string(),
  /** Absolute path to `<askHome>/github/<host>/<owner>/<repo>/<ref>/`. */
  checkoutDir: z.string(),
  npmPackageName: z.string().nullable(),
})

export type SrcModel = z.infer<typeof SrcModelSchema>

export interface RunSrcDeps {
  ensureCheckout?: typeof defaultEnsureCheckout
  log?: (msg: string) => void
  error?: (msg: string) => void
  exit?: (code: number) => void
}

/**
 * Implementation of `ask src <spec>`. Resolves the spec via the shared
 * `ensureCheckout` helper, fetching on cache miss unless `noFetch` is
 * set, then prints the absolute checkout path to stdout.
 *
 * Errors (NoCacheError, resolver failures, unsupported ecosystems) are
 * routed through the injected `error` and `exit` callbacks so the
 * function stays unit-testable without process exits or console writes.
 */
export async function runSrc(options: RunSrcOptions, deps: RunSrcDeps = {}): Promise<void> {
  const ensureCheckout = deps.ensureCheckout ?? defaultEnsureCheckout
  const log = deps.log ?? ((msg: string) => process.stdout.write(`${msg}\n`))
  const error = deps.error ?? ((msg: string) => process.stderr.write(`${msg}\n`))
  const exit = deps.exit ?? ((code: number) => process.exit(code))
  try {
    const result = await ensureCheckout({
      spec: options.spec,
      projectDir: options.projectDir,
      noFetch: options.noFetch,
    })
    if (options.json) {
      const model: SrcModel = {
        spec: options.spec,
        owner: result.owner,
        repo: result.repo,
        ref: result.ref,
        resolvedVersion: result.resolvedVersion,
        checkoutDir: result.checkoutDir,
        npmPackageName: result.npmPackageName ?? null,
      }
      log(JSON.stringify(SrcModelSchema.parse(model), null, 2))
    }
    else {
      log(result.checkoutDir)
    }
  }
  catch (err) {
    if (err instanceof NoCacheError) {
      error(err.message)
      exit(1)
      return
    }
    const message = err instanceof Error ? err.message : String(err)
    error(message)
    exit(1)
  }
}

/**
 * Citty command surface for `ask src`. Wires the citty argument parser
 * to `runSrc` and uses real stdout/stderr/process.exit. Tests should
 * call `runSrc` directly with mocked deps; this wrapper is only for the
 * CLI entrypoint.
 */
export const srcCmd = defineCommand({
  meta: {
    name: 'src',
    description: 'Print the absolute path to a cached library source tree (lazy fetch on cache miss)',
  },
  args: {
    'spec': {
      type: 'positional',
      description: 'Library spec (e.g. react, npm:react@18.2.0, github:facebook/react@v18.2.0)',
      required: true,
    },
    'no-fetch': {
      type: 'boolean',
      description: 'Return cache hit only — exit 1 on cache miss',
    },
    'json': {
      type: 'boolean',
      description: 'Emit resolution as JSON matching SrcModelSchema (spec, owner, repo, ref, resolvedVersion, checkoutDir, npmPackageName)',
    },
  },
  async run({ args }) {
    await runSrc({
      spec: args.spec,
      projectDir: process.cwd(),
      noFetch: Boolean(args['no-fetch']),
      json: Boolean(args.json),
    })
  },
})

File: crates/ask/src/commands/src.rs
//! `ask src <spec>` — print the absolute path to a cached library source tree,
//! fetching on cache miss unless `--no-fetch`. Rust port of `commands/src.ts`.
//!
//! `run_src` returns the exact stdout string on success and an error on failure;
//! the thin CLI wrapper prints it and maps any error to a non-zero exit. This
//! keeps the resolution logic unit-testable without process exits or real I/O.

use std::path::PathBuf;

use anyhow::Result;
use serde::Serialize;

use crate::commands::ensure_checkout::{
    ensure_checkout, EnsureCheckoutDeps, EnsureCheckoutOptions,
};
use crate::http::HttpClient;

/// Options for [`run_src`].
#[derive(Debug, Clone)]
pub struct RunSrcOptions {
    pub spec: String,
    pub project_dir: PathBuf,
    pub no_fetch: bool,
    pub json: bool,
}

/// `ask src <spec> --json` output — the stable machine-readable handoff that
/// downstream tools (e.g. csp) consume. `checkout_dir` is the version-pinned,
/// content-stable store path they index. Mirrors `SrcModelSchema`.
#[derive(Debug, Serialize, PartialEq, Eq)]
pub struct SrcModel {
    pub spec: String,
    pub owner: String,
    pub repo: String,
    #[serde(rename = "ref")]
    pub reference: String,
    #[serde(rename = "resolvedVersion")]
    pub resolved_version: String,
    #[serde(rename = "checkoutDir")]
    pub checkout_dir: String,
    #[serde(rename = "npmPackageName")]
    pub npm_package_name: Option<String>,
}

/// Resolve `spec` via the shared checkout helper and return the text to print:
/// the checkout path (default) or a JSON [`SrcModel`] (`json`). On a
/// `NoCacheError` or any resolver failure this returns an error whose message
/// the CLI writes to stderr before exiting non-zero — matching the TS behaviour
/// where both cases route through `error()` + `exit(1)`.
pub fn run_src(client: &dyn HttpClient, options: &RunSrcOptions) -> Result<String> {
    run_src_with(client, options, &EnsureCheckoutDeps::default())
}

/// [`run_src`] with injectable checkout deps (tests pass an offline fetcher /
/// overridden ASK home).
pub fn run_src_with(
    client: &dyn HttpClient,
    options: &RunSrcOptions,
    deps: &EnsureCheckoutDeps,
) -> Result<String> {
    let result = ensure_checkout(
        client,
        &EnsureCheckoutOptions {
            spec: options.spec.clone(),
            project_dir: options.project_dir.clone(),
            no_fetch: options.no_fetch,
        },
        deps,
    )?;

    if options.json {
        let model = SrcModel {
            spec: options.spec.clone(),
            owner: result.owner,
            repo: result.repo,
            reference: result.reference,
            resolved_version: result.resolved_version,
            checkout_dir: result.checkout_dir.to_string_lossy().into_owned(),
            npm_package_name: result.npm_package_name,
        };
        Ok(serde_json::to_string_pretty(&model)?)
    } else {
        Ok(result.checkout_dir.to_string_lossy().into_owned())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::ensure_checkout::NoCacheError;
    use crate::http::mock::MockClient;
    use crate::store::github_store_path;

    fn deps_home(home: &std::path::Path) -> EnsureCheckoutDeps<'static> {
        EnsureCheckoutDeps {
            ask_home: Some(home.to_path_buf()),
            fetcher: None,
        }
    }

    #[test]
    fn prints_checkout_dir_on_cache_hit() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let dir =
            github_store_path(home.path(), "github.com", "facebook", "react", "v18.2.0").unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        let out = run_src_with(
            &MockClient::new(),
            &RunSrcOptions {
                spec: "github:facebook/react@v18.2.0".into(),
                project_dir: proj.path().to_path_buf(),
                no_fetch: false,
                json: false,
            },
            &deps_home(home.path()),
        )
        .unwrap();
        assert_eq!(out, dir.to_string_lossy());
    }

    #[test]
    fn json_output_matches_schema_field_names() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let dir =
            github_store_path(home.path(), "github.com", "facebook", "react", "v18.2.0").unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        let out = run_src_with(
            &MockClient::new(),
            &RunSrcOptions {
                spec: "github:facebook/react@v18.2.0".into(),
                project_dir: proj.path().to_path_buf(),
                no_fetch: false,
                json: true,
            },
            &deps_home(home.path()),
        )
        .unwrap();
        assert!(out.contains(r#""ref": "v18.2.0""#));
        assert!(out.contains(r#""resolvedVersion": "v18.2.0""#));
        assert!(out.contains(r#""npmPackageName": null"#));
        assert!(out.contains(r#""checkoutDir""#));
    }

    #[test]
    fn no_cache_error_propagates() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let err = run_src_with(
            &MockClient::new(),
            &RunSrcOptions {
                spec: "github:owner/repo@v1.2.3".into(),
                project_dir: proj.path().to_path_buf(),
                no_fetch: true,
                json: false,
            },
            &deps_home(home.path()),
        )
        .unwrap_err();
        assert!(err.downcast_ref::<NoCacheError>().is_some());
    }
}
